crates/swc/src/dropped_comments_preserver.rs

```diff
@@ -0,0 +1,135 @@
+        script.visit_mut_children_with(self);
+        self.shift_comments_to_known_spans();
+    }
+
+    fn visit_mut_span(&mut self, span: &mut Span) {
+        if span.is_dummy() || self.is_first_span {
+            self.is_first_span = false;
+            return;
+        }
+
+        self.known_spans.push(*span);
+        span.visit_mut_children_with(self)
+    }
+}
+
+impl DroppedCommentsPreserver<'_> {
+    fn shift_comments_to_known_spans(&mut self) {
+        if let Some(comments) = self.comments {
+            let trailing_comments = self.shift_leading_comments(comments);
+
+            self.shift_trailing_comments(trailing_comments);
+        }
+    }
+
+    /// We'll be shifting all comments to known span positions, so drain the
+    /// current comments first to limit the amount of look ups needed into
+    /// the hashmaps.
+    ///
+    /// This way, we only need to take the comments once, and then add them back
+    /// once.
+    fn collect_existing_comments(&mut self, comments: &SingleThreadedComments) -> CommentEntries {
+        let (mut leading_comments, mut trailing_comments) = comments.borrow_all_mut();
+        let mut existing_comments: CommentEntries = leading_comments
+            .drain()
+            .chain(trailing_comments.drain())
+            .collect();
+
+        existing_comments.sort_by(|(bp_a, _), (bp_b, _)| bp_a.cmp(bp_b));
+
+        existing_comments
+    }
+
+    /// Shift all comments to known leading positions.
+    /// This prevents trailing comments from ending up associated with
+    /// nodes that will not emit trailing comments, while
+    /// preserving any comments that might show up after all code positions.
+    ///
+    /// This maintains the highest fidelity between existing comment positions
+    /// of pre and post compiled code.
+    fn shift_leading_comments(&mut self, comments: &SingleThreadedComments) -> CommentEntries {
+        let mut existing_comments = self.collect_existing_comments(comments);
+
+        for span in self.known_spans.iter() {
+            let (comments_to_move, next_byte_positions): (CommentEntries, CommentEntries) =
+                existing_comments
+                    .drain(..)
+                    .partition(|(bp, _)| bp <= &span.lo);
+
+            existing_comments.extend(next_byte_positions);
+
+            let collected_comments = comments_to_move.into_iter().flat_map(|(_, c)| c).collect();
+
+            self.comments
+                .add_leading_comments(span.lo, collected_comments)
+        }
+
+        existing_comments
+    }
+
+    /// These comments trail all known span lo byte positions.
+    /// Therefore, by shifting them to trail the highest known hi position, we
+    /// ensure that any remaining trailing comments are emitted in a
+    /// similar location
+    fn shift_trailing_comments(&mut self, remaining_comment_entries: CommentEntries) {
+        let last_trailing =
+            self.known_spans.iter().fold(
```

Comment:
Please use `.iter().copied()` instead.
`Span` is `Copy`, so we don't need indirection here.